### PR TITLE
Hook performance

### PR DESCRIPTION
**Pull request by B-head**
Improve run-time performance.

Switch to a lightweight debughook at runtime. At the beginning of the hook, check the breakpoint and current line to reduce the if statement being executed.

### CODE REVIEW

**Comment by tomblind** [comment]
This is a big change and it will need careful testing. It may be awhile before I have the time to do that, but I'll get to it when I can.

File: debugger/debugger.ts [hunk 1]
```diff
@@ -574,45 +581,37 @@ export namespace Debugger {
         return false;
     }
 
-    function debugHook(event: "call" | "return" | "tail return" | "count" | "line", line?: number) {
-        const stackOffset = 2;
+    const topFrameStackOffset = 2;
 
+    function isIgnoreStepBreak(topFrame: debug.FunctionInfo): boolean {
```
Comment (tomblind) [inline debugger/debugger.ts:586]:
```suggestion
    function ignoreStepBreak(topFrame: debug.FunctionInfo): boolean {
```

File: debugger/debugger.ts [hunk 2]
```diff
@@ -574,45 +581,37 @@ export namespace Debugger {
...
 
         //Ignore builtin lua functions (luajit)
         if (topFrame.short_src && topFrame.short_src.sub(1, builtinFunctionPrefix.length) === builtinFunctionPrefix) {
+            return true;
+        }
+
+        return false;
+    }
+
+    function checkAllBreakPoint(stackOffset: number, line: number) {
```
Comment (tomblind) [inline debugger/debugger.ts:600]:
```suggestion
    function checkAllBreakPoints(stackOffset: number, line: number) {
```

File: debugger/debugger.ts [hunk 3]
```diff
@@ -758,6 +785,24 @@ export namespace Debugger {
         }
     }
 
+    function setHook(hook?: debug.Hook) {
+        if (!!hook) {
```
Comment (tomblind) [inline debugger/debugger.ts:789]:
```suggestion
        if (hook !== undefined) {
```

File: debugger/debugger.ts [hunk 4]
```diff
@@ -758,6 +785,24 @@ export namespace Debugger {
...
+            debug.sethook(hook, "l");
+        } else {
+            debug.sethook();
+        }
+
+        for (const [thread] of pairs(threadIds)) {
+            if (isThread(thread) && coroutine.status(thread) !== "dead") {
+                if (!!hook) {
```
Comment (tomblind) [inline debugger/debugger.ts:797]:
```suggestion
                if (hook !== undefined) {
```

File: debugger/debugger.ts [hunk 5]
```diff
@@ -814,6 +847,10 @@ export namespace Debugger {
 
     export function triggerBreak() {
         breakAtDepth = math.huge;
+        breakInThread = undefined;
+        if (hookStack.length > 0) {
```
Comment (tomblind) [inline debugger/debugger.ts:851]: If `triggerBreak` is called before `debugFunction`, `runHook` will be set the debugger won't break immediately as expected.

File: debugger/breakpoint.ts [hunk 6]
```diff
@@ -39,6 +39,14 @@ export namespace Breakpoint {
         return current;
     }
 
+    export function getLines(): {[key: number]: boolean} {
+        const ret: {[key: number]: boolean} = {};
```
Comment (tomblind) [inline debugger/breakpoint.ts:43]:
```suggestion
        const lines: {[key: number]: boolean} = {};
```

**Comment by B-head** [comment]
@tomblind  Thanks for the review. I have committed to all of your suggestions.

**Pull request closed without merging**